### PR TITLE
[WIP] Deploy tool arch

### PR DESCRIPTION
**Pull request by mjudeikis**
With lack of access to the wiki, putting this here for now for discussion. Will move when my access is restored.

@JackQuincy @jim-minter @julienstroheker @amanohar 

related to: https://github.com/Azure/ARO-RP/pull/233

---
This change is [Reviewable](https://reviewable.io/reviews/azure/aro-rp/252)

### CODE REVIEW

File: docs/arch/deploy-upgrade-tool.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+-   Idempotent
+-   Configuration for deployment
+-   Automated secret rotation (not now)
+
+**Proposal**:
+Command line tool:
+`./aro deploy {region}`
+Generic configuration file config.yaml:
+```yaml
+rps:
```
[the remark this answers is not in the document]
Comment (mjudeikis) [reply docs/arch/deploy-upgrade-tool.md]: I was thinking to keep it flat and higher-level cloud would be configured via overrides. As it should be possible to configure deployment using `configuration` stanza. So override. Maybe adding `name` field for clarification.

File: docs/arch/deploy-upgrade-tool.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  sshPublicKey:
+  vmssName:
+```
+
+**Flow**:
+
+-   Each region can have substructure of same configuration struct as root. Individual regional configuration structs act as overrides for global configurable
+-   For the start configuration sub-structure will be YAML representation of ARM parameters[1] files, used at the moment.
+
+Install:
```
[the remark this answers is not in the document]
Comment (mjudeikis) [reply docs/arch/deploy-upgrade-tool.md]: Breakpoint was "by default" in my head. Will update it there to say it outloud.

File: docs/arch/deploy-upgrade-tool.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	-   Create Vault
+Deploy:
+	-   Setup encryption keys
+	-   Submit a request for Certificates to be generated and wait for certificates
+	 -   Create new certificates and wait
+	-   Create A and NS global records for RP
+	-   Initiate ACR new replica (optional for now)
+	-   Deploy RP infrastructure
+	-   Wait for RP to come online
+	-   Deploy external azure integration components (optional for now)
```
[the remark this answers is not in the document]
Comment (mjudeikis) [reply docs/arch/deploy-upgrade-tool.md]: This is a placeholder in the document for any future integration we might have. In example ARC, or other managment compnents.

File: docs/arch/deploy-upgrade-tool.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+Upgrade
+	-   If required run PreDeploy (configured via env hook)
+	-   Request new certificates for rotation (optional for now, requires RP core code changes)
```
[the remark this answers is not in the document]
Comment (mjudeikis) [reply docs/arch/deploy-upgrade-tool.md]: Rotation included?

File: docs/arch/deploy-upgrade-tool.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+rps:
+ - location: eastus
+    subscriptionId: <subscriptionID>
+    resourceGroupName: <ResourceGroupName>
+ - location: westus
+    subscriptionId: <subscriptionID>
+    resourceGroupName: <ResourceGroupName>
+    configuration:
+       <same structure as global configuration>
+configuration:
```
[the remark this answers is not in the document]
Comment (mjudeikis) [reply docs/arch/deploy-upgrade-tool.md]: Yes, but I think at the moment they should stay until we get comfortable with the process

File: docs/arch/deploy-upgrade-tool.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  domainName:
+  extraCosmosDBIPs:
+  extraKeyvaultAccessPolicies:
+  fpServicePrincipalId:
+  mdmFrontendUrl :
+  mdsdConfigVersion:
+  mdsdEnvironment:
+  pullSecret:
+  rpImage:
+  rpImageAuth:
```
[the remark this answers is not in the document]
Comment (mjudeikis) [reply docs/arch/deploy-upgrade-tool.md]: I think this is out of scope for this for now :/

**Comment by jim-minter** [comment]
have put this in wiki

**Pull request closed without merging**